src/main/java/org/mule/extension/validation/internal/validator/SizeValidator.java

```diff
@@ -8,68 +8,71 @@
+import org.mule.runtime.api.el.BindingContext;
+import org.mule.runtime.api.el.ExpressionLanguage;
 import org.mule.runtime.api.i18n.I18nMessage;
+import org.mule.runtime.api.metadata.TypedValue;
 
 import java.util.Collection;
 import java.util.Map;
 
-import org.apache.commons.lang3.ArrayUtils;
-
 /**
- * An {@link AbstractValidator} which verifies that {@link #value} has a size between certain inclusive boundaries. This validator
- * is capable of handling instances of {@link String}, {@link Collection}, {@link Map} and arrays
+ * An {@link AbstractValidator} which verifies that {@link #typedValue} has a size between certain inclusive boundaries. This
+ * validator is capable of handling instances of {@link String}, {@link Collection}, {@link Map} and arrays
  *
  * @since 1.0
  */
 public class SizeValidator extends AbstractValidator {
 
-  private final Object value;
+  private final TypedValue typedValue;
   private final int minSize;
   private final Integer maxSize;
+  private final ExpressionLanguage expressionLanguage;
 
   private I18nMessage errorMessage;
 
-  public SizeValidator(Object value, int minSize, Integer maxSize, ValidationContext validationContext) {
+  public SizeValidator(TypedValue typedValue, int minSize, Integer maxSize, ValidationContext validationContext,
+                       ExpressionLanguage expressionLanguage) {
     super(validationContext);
-    this.value = value;
+    this.typedValue = typedValue;
     this.minSize = minSize;
     this.maxSize = maxSize;
+    this.expressionLanguage = expressionLanguage;
   }
 
   @Override
   public ValidationResult validate() {
-    int inputLength = getSize(value);
+    int inputLength = getSize(typedValue);
     if (inputLength < minSize) {
-      errorMessage = getMessages().lowerThanMinSize(value, minSize, inputLength);
+      errorMessage = getMessages().lowerThanMinSize(typedValue.getValue(), minSize, inputLength);
       return fail();
     }
 
     if (maxSize != null && inputLength > maxSize) {
-      errorMessage = getMessages().greaterThanMaxSize(value, maxSize, inputLength);
+      errorMessage = getMessages().greaterThanMaxSize(typedValue.getValue(), maxSize, inputLength);
       return fail();
     }
 
     return ok();
   }
 
-  private int getSize(Object value) {
-    checkArgument(value != null, "Cannot check size of a null value");
-    if (value instanceof String) {
-      return ((String) value).length();
-    } else if (value instanceof Collection) {
-      return ((Collection<?>) value).size();
-    } else if (value instanceof Map) {
-      return ((Map<?, ?>) value).size();
-    } else if (value.getClass().isArray()) {
-      return ArrayUtils.getLength(value);
+  private int getSize(TypedValue typedValue) {
+    checkArgument(typedValue.getValue() != null, "Cannot check size of a null value");
+
+    BindingContext context = BindingContext.builder().addBinding("payload", typedValue).build();
```

Comment:
"payload" to constant

Reply:
we should first check for known elements instead of always evaluating an expression. List, string, map, those all have a size that can be obtained easily. This would improve performance